<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial • metacal</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metacal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial.html">Tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikemc/metacal">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mikemc/metacal/blob/master/vignettes/tutorial.Rmd"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>tutorial.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co">#&gt; Registered S3 methods overwritten by 'ggplot2':</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co">#&gt;   method         from </span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">#&gt;   [.quosures     rlang</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co">#&gt;   c.quosures     rlang</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#&gt;   print.quosures rlang</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co">#&gt; ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──</span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">#&gt; ✔ ggplot2 3.1.1     ✔ purrr   0.3.2</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co">#&gt; ✔ tibble  2.1.3     ✔ dplyr   0.8.3</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co">#&gt; ✔ tidyr   0.8.3     ✔ stringr 1.4.0</span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="co">#&gt; ✔ readr   1.3.1     ✔ forcats 0.4.0</span></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="co">#&gt; ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(metacal)</span></code></pre></div>
<p>This tutorial goes through the process of estimating bias and performing calibration in the special case where the bias of all the taxa of interest can be directly measured from the control samples. It covers estimating the bias with associated uncertainty from a set of control samples, evaluating the consistency of bias across multiple controls, and using the bias estimate to calibrate samples with unknown composition.</p>
<div id="load-the-demonstration-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-demonstration-dataset" class="anchor"></a>Load the demonstration dataset</h1>
<p>For this tutorial, we will use the data from the cell-mixtures experiment of Brooks et al. (2015), which is included in this package. Each sample is a cellular mock community consisting of an even mixture of one to seven of a set of seven bacterial species. The sample metadata and the observed and actual species abundances are stored as <code>.csv</code> files. The abundances are stored in standard OTU-table format with samples as rows and taxa as columns.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>sam &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"brooks2015-sample-data.csv"</span>, </span>
<span id="cb2-2"><a href="#cb2-2"></a>        <span class="dt">package =</span> <span class="st">"metacal"</span>))</span>
<span id="cb2-3"><a href="#cb2-3"></a>observed &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"brooks2015-observed.csv"</span>, </span>
<span id="cb2-4"><a href="#cb2-4"></a>        <span class="dt">package =</span> <span class="st">"metacal"</span>))</span>
<span id="cb2-5"><a href="#cb2-5"></a>actual &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"brooks2015-actual.csv"</span>, </span>
<span id="cb2-6"><a href="#cb2-6"></a>        <span class="dt">package =</span> <span class="st">"metacal"</span>))</span>
<span id="cb2-7"><a href="#cb2-7"></a>sam</span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt; # A tibble: 80 x 6</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt;    Sample Plate Barcode Mixture_type Num_species Species_list              </span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#&gt;    &lt;chr&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;                     </span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt;  1 s1-1       1       1 Cells                  3 Lactobacillus_crispatus;P…</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt;  2 s1-2       1       2 Cells                  3 Gardnerella_vaginalis;Ato…</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt;  3 s1-3       1       3 Cells                  3 Gardnerella_vaginalis;Ato…</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt;  4 s1-4       1       4 Cells                  3 Atopobium_vaginae;Lactoba…</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt;  5 s1-5       1       5 Cells                  2 Gardnerella_vaginalis;Pre…</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt;  6 s1-6       1       6 Cells                  3 Atopobium_vaginae;Lactoba…</span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co">#&gt;  7 s1-7       1       7 Cells                  2 Atopobium_vaginae;Strepto…</span></span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="co">#&gt;  8 s1-8       1       8 Cells                  2 Gardnerella_vaginalis;Str…</span></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co">#&gt;  9 s1-9       1       9 Cells                  3 Atopobium_vaginae;Sneathi…</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="co">#&gt; 10 s1-10      1      10 Cells                  3 Atopobium_vaginae;Lactoba…</span></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co">#&gt; # … with 70 more rows</span></span>
<span id="cb2-22"><a href="#cb2-22"></a>observed</span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="co">#&gt; # A tibble: 80 x 9</span></span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="co">#&gt;    Sample Atopobium_vagin… Gardnerella_vag… Lactobacillus_c…</span></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span></span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="co">#&gt;  1 s1-1                  1                1            13670</span></span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="co">#&gt;  2 s1-10              1028                0            10310</span></span>
<span id="cb2-28"><a href="#cb2-28"></a><span class="co">#&gt;  3 s1-11                 0              181             3956</span></span>
<span id="cb2-29"><a href="#cb2-29"></a><span class="co">#&gt;  4 s1-12              1424                0                0</span></span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="co">#&gt;  5 s1-13                 0                0             3199</span></span>
<span id="cb2-31"><a href="#cb2-31"></a><span class="co">#&gt;  6 s1-14                 0                0             9348</span></span>
<span id="cb2-32"><a href="#cb2-32"></a><span class="co">#&gt;  7 s1-15               572                0                1</span></span>
<span id="cb2-33"><a href="#cb2-33"></a><span class="co">#&gt;  8 s1-16                 0                0                1</span></span>
<span id="cb2-34"><a href="#cb2-34"></a><span class="co">#&gt;  9 s1-17                 2                1               19</span></span>
<span id="cb2-35"><a href="#cb2-35"></a><span class="co">#&gt; 10 s1-18                 0              865             9892</span></span>
<span id="cb2-36"><a href="#cb2-36"></a><span class="co">#&gt; # … with 70 more rows, and 5 more variables: Lactobacillus_iners &lt;dbl&gt;,</span></span>
<span id="cb2-37"><a href="#cb2-37"></a><span class="co">#&gt; #   Prevotella_bivia &lt;dbl&gt;, Sneathia_amnii &lt;dbl&gt;,</span></span>
<span id="cb2-38"><a href="#cb2-38"></a><span class="co">#&gt; #   Streptococcus_agalactiae &lt;dbl&gt;, Other &lt;dbl&gt;</span></span>
<span id="cb2-39"><a href="#cb2-39"></a>actual</span>
<span id="cb2-40"><a href="#cb2-40"></a><span class="co">#&gt; # A tibble: 80 x 8</span></span>
<span id="cb2-41"><a href="#cb2-41"></a><span class="co">#&gt;    Sample Atopobium_vagin… Gardnerella_vag… Lactobacillus_c…</span></span>
<span id="cb2-42"><a href="#cb2-42"></a><span class="co">#&gt;    &lt;chr&gt;             &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span></span>
<span id="cb2-43"><a href="#cb2-43"></a><span class="co">#&gt;  1 s1-1              0                0                0.333</span></span>
<span id="cb2-44"><a href="#cb2-44"></a><span class="co">#&gt;  2 s1-10             0.333            0                0.333</span></span>
<span id="cb2-45"><a href="#cb2-45"></a><span class="co">#&gt;  3 s1-11             0                0.333            0.333</span></span>
<span id="cb2-46"><a href="#cb2-46"></a><span class="co">#&gt;  4 s1-12             0.5              0                0    </span></span>
<span id="cb2-47"><a href="#cb2-47"></a><span class="co">#&gt;  5 s1-13             0                0                0.333</span></span>
<span id="cb2-48"><a href="#cb2-48"></a><span class="co">#&gt;  6 s1-14             0                0                0.5  </span></span>
<span id="cb2-49"><a href="#cb2-49"></a><span class="co">#&gt;  7 s1-15             0.333            0                0    </span></span>
<span id="cb2-50"><a href="#cb2-50"></a><span class="co">#&gt;  8 s1-16             0                0                0    </span></span>
<span id="cb2-51"><a href="#cb2-51"></a><span class="co">#&gt;  9 s1-17             0                0                0    </span></span>
<span id="cb2-52"><a href="#cb2-52"></a><span class="co">#&gt; 10 s1-18             0                0.5              0.5  </span></span>
<span id="cb2-53"><a href="#cb2-53"></a><span class="co">#&gt; # … with 70 more rows, and 4 more variables: Lactobacillus_iners &lt;dbl&gt;,</span></span>
<span id="cb2-54"><a href="#cb2-54"></a><span class="co">#&gt; #   Prevotella_bivia &lt;dbl&gt;, Sneathia_amnii &lt;dbl&gt;,</span></span>
<span id="cb2-55"><a href="#cb2-55"></a><span class="co">#&gt; #   Streptococcus_agalactiae &lt;dbl&gt;</span></span></code></pre></div>
<p>The <code>Other</code> column in <code>observed</code> includes all reads assigned to other species due to contamination or taxonomic misassignment.</p>
<p>The samples contain mixtures of 1, 2, 3, 4, or 7 species,</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>sam <span class="op">%&gt;%</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="st">    </span><span class="kw">group_by</span>(Num_species) <span class="op">%&gt;%</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="st">    </span>count</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; # A tibble: 5 x 2</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; # Groups:   Num_species [5]</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt;   Num_species     n</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt;         &lt;dbl&gt; &lt;int&gt;</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; 1           1     9</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; 2           2    25</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt; 3           3    43</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt; 4           4     1</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt; 5           7     2</span></span></code></pre></div>
<p>For this tutorial, we will pick the 4-species sample and two each of the 2-, 3-, and 7-species samples to serve as the control samples that we will use to estimate bias.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>controls &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"s1-23"</span>, <span class="st">"s1-35"</span>, <span class="st">"s2-4"</span>, <span class="st">"s2-5"</span>, <span class="st">"s2-8"</span>, <span class="st">"s2-9"</span>, <span class="st">"s2-14"</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a>sam <span class="op">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Sample <span class="op">%in%</span><span class="st"> </span>controls) <span class="op">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="st">    </span><span class="kw">arrange</span>(Num_species)</span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt; # A tibble: 7 x 6</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">#&gt;   Sample Plate Barcode Mixture_type Num_species Species_list               </span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;                      </span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">#&gt; 1 s2-4       2       4 Cells                  2 Atopobium_vaginae;Sneathia…</span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">#&gt; 2 s2-5       2       5 Cells                  2 Atopobium_vaginae;Lactobac…</span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">#&gt; 3 s2-8       2       8 Cells                  3 Gardnerella_vaginalis;Lact…</span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">#&gt; 4 s2-9       2       9 Cells                  3 Gardnerella_vaginalis;Lact…</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">#&gt; 5 s1-35      1      35 Cells                  4 Gardnerella_vaginalis;Lact…</span></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co">#&gt; 6 s1-23      1      23 Cells                  7 Gardnerella_vaginalis;Atop…</span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co">#&gt; 7 s2-14      2      14 Cells                  7 Gardnerella_vaginalis;Atop…</span></span>
<span id="cb4-15"><a href="#cb4-15"></a>sam &lt;-<span class="st"> </span>sam <span class="op">%&gt;%</span></span>
<span id="cb4-16"><a href="#cb4-16"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Type =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(Sample <span class="op">%in%</span><span class="st"> </span>controls, <span class="st">"Control"</span>, <span class="st">""</span>))</span>
<span id="cb4-17"><a href="#cb4-17"></a>actual &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(actual, Sample <span class="op">%in%</span><span class="st"> </span>controls)</span></code></pre></div>
<p>We’ll also abbreviate the taxa names for greater compactness in R output,</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>tax_abbrev &lt;-<span class="st"> </span><span class="cf">function</span> (taxa) {</span>
<span id="cb5-2"><a href="#cb5-2"></a>    m &lt;-<span class="st"> </span><span class="kw">str_match</span>(taxa, <span class="st">"([A-Z])[a-z]+_([a-z]{2})[a-z]*"</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(m[,<span class="dv">1</span>]), taxa, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(m[,<span class="dv">2</span>], m[,<span class="dv">3</span>]))</span>
<span id="cb5-4"><a href="#cb5-4"></a>}</span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(observed)</span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">#&gt; [1] "Sample"                   "Atopobium_vaginae"       </span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co">#&gt; [3] "Gardnerella_vaginalis"    "Lactobacillus_crispatus" </span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co">#&gt; [5] "Lactobacillus_iners"      "Prevotella_bivia"        </span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="co">#&gt; [7] "Sneathia_amnii"           "Streptococcus_agalactiae"</span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="co">#&gt; [9] "Other"</span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="kw">tax_abbrev</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(observed)[<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="co">#&gt; [1] "Ava"   "Gva"   "Lcr"   "Lin"   "Pbi"   "Sam"   "Sag"   "Other"</span></span>
<span id="cb5-13"><a href="#cb5-13"></a>observed &lt;-<span class="st"> </span><span class="kw">rename_at</span>(observed, <span class="kw">vars</span>(<span class="op">-</span>Sample), tax_abbrev)</span>
<span id="cb5-14"><a href="#cb5-14"></a>actual &lt;-<span class="st"> </span><span class="kw">rename_at</span>(actual, <span class="kw">vars</span>(<span class="op">-</span>Sample), tax_abbrev)</span></code></pre></div>
<p><strong>Note:</strong> Bias estimation requires that the <code>observed</code> and <code>actual</code> tables use the same taxonomic identifiers, so that we can match the observed and true abundances for each taxon. We’ve already taken care of this in the demonstration data. Real applications may require an extra step to reconcile the taxonomic identifiers output by the bioinformatics pipeline with those in the table of actual compositions.</p>
</div>
<div id="estimate-bias" class="section level1">
<h1 class="hasAnchor">
<a href="#estimate-bias" class="anchor"></a>Estimate bias</h1>
<div id="compute-the-error-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-the-error-matrix" class="anchor"></a>Compute the error matrix</h2>
<p>To estimate bias, we first need to get a matrix with the compositional error between the observed and actual compositions for the control samples. This matrix should have samples corresponding to rows, taxa corresponding to columns, and columns named by the taxa.</p>
<div id="approach-1-matrices" class="section level3">
<h3 class="hasAnchor">
<a href="#approach-1-matrices" class="anchor"></a>Approach 1: Matrices</h3>
<p>First, convert to numeric matrices using the <code><a href="../reference/as_matrix.html">as_matrix()</a></code> helper function,</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>observed_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_matrix.html">as_matrix</a></span>(observed, <span class="dt">rownames =</span> Sample)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(observed_mat)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt;        Ava Gva   Lcr  Lin   Pbi   Sam  Sag Other</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; s1-1     1   1 13670    0  7544     1 1506     7</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; s1-10 1028   0 10310    2     1 14947    2     3</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">#&gt; s1-11    0 181  3956 7598     6     2    0     2</span></span>
<span id="cb6-7"><a href="#cb6-7"></a><span class="co">#&gt; s1-12 1424   0     0    1 21708     7    0     8</span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co">#&gt; s1-13    0   0  3199    0  1854  6501    0     3</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="co">#&gt; s1-14    0   0  9348    2     0     3 1475    10</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>actual_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_matrix.html">as_matrix</a></span>(actual, <span class="dt">rownames =</span> Sample)</span>
<span id="cb6-11"><a href="#cb6-11"></a>actual_mat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="dv">3</span>)</span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="co">#&gt;         Ava   Gva   Lcr   Lin   Pbi   Sam   Sag</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co">#&gt; s1-23 0.143 0.143 0.143 0.143 0.143 0.143 0.143</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co">#&gt; s1-35 0.000 0.250 0.000 0.250 0.250 0.250 0.000</span></span>
<span id="cb6-15"><a href="#cb6-15"></a><span class="co">#&gt; s2-14 0.143 0.143 0.143 0.143 0.143 0.143 0.143</span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="co">#&gt; s2-4  0.500 0.000 0.000 0.000 0.000 0.500 0.000</span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="co">#&gt; s2-5  0.500 0.000 0.000 0.500 0.000 0.000 0.000</span></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="co">#&gt; s2-8  0.000 0.333 0.000 0.333 0.333 0.000 0.000</span></span>
<span id="cb6-19"><a href="#cb6-19"></a><span class="co">#&gt; s2-9  0.000 0.333 0.333 0.000 0.333 0.000 0.000</span></span></code></pre></div>
<p>Then, get compatible matrices with just the control samples and taxa, making sure that rows and columns are in the same order,</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>control_samples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(actual_mat)</span>
<span id="cb7-2"><a href="#cb7-2"></a>control_taxa &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(actual_mat)</span>
<span id="cb7-3"><a href="#cb7-3"></a>observed_mat &lt;-<span class="st"> </span>observed_mat[control_samples, control_taxa]</span>
<span id="cb7-4"><a href="#cb7-4"></a>actual_mat &lt;-<span class="st"> </span>actual_mat[control_samples, control_taxa]</span></code></pre></div>
<p>The final step is to divide the observed by the actual abundances to get the error matrix,</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>error_mat &lt;-<span class="st"> </span>observed_mat <span class="op">/</span><span class="st"> </span>actual_mat</span>
<span id="cb8-2"><a href="#cb8-2"></a>error_mat</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">#&gt;        Ava  Gva   Lcr   Lin   Pbi   Sam  Sag</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; s1-23 2457 1421 17241 34202 10220 33586 1967</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#&gt; s1-35  Inf 1068   NaN 33316 12760 21780  Inf</span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">#&gt; s2-14 3577 2156 15792 29225 11536 32949 1813</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; s2-4  2898  NaN   Inf   Inf   Inf 41272  NaN</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#&gt; s2-5  1766  NaN   NaN 28424   Inf   Inf  NaN</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co">#&gt; s2-8   NaN 2238   Inf 61653 23298   Inf  NaN</span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">#&gt; s2-9   NaN 1473 29616   NaN 19230   NaN  NaN</span></span></code></pre></div>
<p>This step can lead to two types of problems. Infinite values (<code>Inf</code>s) arise from taxa that have a positive abundance in <code>observed</code> but not in <code>actual</code>. We see several cases of that here. The opposite situation, a taxon with a positive abundance in <code>actual</code> but an abundance of 0 in <code>observed</code>, occurs when taxa present in the controls were not detected. This situation does not occur here but is likely to occur in control samples with low-frequency or low-efficiency taxa and low sequencing depths. In contrast, the <code>NaN</code>s we see are fine: these are caused by dividing <code>0/0</code> and reflect taxa that are neither observed nor actually present in the control. To ensure that we have no <code>Inf</code>s or <code>0</code>s in the error matrix, we can add a pseudocount (or apply another zero-removal strategy) to the observed matrix and “mask” (set to 0) the observations of taxa not in <code>actual</code>,</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>observed_mat0 &lt;-<span class="st"> </span>(observed_mat <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>) <span class="op">*</span><span class="st"> </span>(actual_mat <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span></code></pre></div>
<p>and then compute the error matrix as before,</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>error_mat &lt;-<span class="st"> </span>observed_mat0 <span class="op">/</span><span class="st"> </span>actual_mat</span>
<span id="cb10-2"><a href="#cb10-2"></a>error_mat</span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#&gt;          Ava    Gva     Lcr     Lin     Pbi     Sam    Sag</span></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co">#&gt; s1-23 2460.5 1424.5 17244.5 34205.5 10223.5 33589.5 1970.5</span></span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#&gt; s1-35    NaN 1070.0     NaN 33318.0 12762.0 21782.0    NaN</span></span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="co">#&gt; s2-14 3580.5 2159.5 15795.5 29228.5 11539.5 32952.5 1816.5</span></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co">#&gt; s2-4  2899.0    NaN     NaN     NaN     NaN 41273.0    NaN</span></span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="co">#&gt; s2-5  1767.0    NaN     NaN 28425.0     NaN     NaN    NaN</span></span>
<span id="cb10-9"><a href="#cb10-9"></a><span class="co">#&gt; s2-8     NaN 2239.5     NaN 61654.5 23299.5     NaN    NaN</span></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co">#&gt; s2-9     NaN 1474.5 29617.5     NaN 19231.5     NaN    NaN</span></span>
<span id="cb10-11"><a href="#cb10-11"></a><span class="co"># Check that all entries either between 0 and Infinity, or NaN</span></span>
<span id="cb10-12"><a href="#cb10-12"></a>error_mat <span class="op">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13"></a><span class="st">    </span>{(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/is.finite">is.finite</a></span>(.) <span class="op">&amp;</span><span class="st"> </span>(. <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)) <span class="op">|</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/is.finite">is.nan</a></span>(.)} <span class="op">%&gt;%</span></span>
<span id="cb10-14"><a href="#cb10-14"></a><span class="st">    </span>all</span>
<span id="cb10-15"><a href="#cb10-15"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Because we treat the abundances and the error as compositional vectors, there is no need to normalize the abundances to proportions before or after computing the error.</p>
</div>
<div id="approach-2-tidy-data-frames" class="section level3">
<h3 class="hasAnchor">
<a href="#approach-2-tidy-data-frames" class="anchor"></a>Approach 2: Tidy data frames</h3>
<p>First, we combine the observed and actual compositions into a single “tidy” data frame, where each row corresponds to an individual (sample, taxon) observation,</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>observed.tidy &lt;-<span class="st"> </span>observed <span class="op">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="st">    </span><span class="kw">gather</span>(<span class="st">"Taxon"</span>, <span class="st">"Observed"</span>, <span class="op">-</span>Sample)</span>
<span id="cb11-3"><a href="#cb11-3"></a>actual.tidy &lt;-<span class="st"> </span>actual <span class="op">%&gt;%</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="st">    </span><span class="kw">gather</span>(<span class="st">"Taxon"</span>, <span class="st">"Actual"</span>, <span class="op">-</span>Sample)</span>
<span id="cb11-5"><a href="#cb11-5"></a>joint &lt;-<span class="st"> </span>actual.tidy <span class="op">%&gt;%</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="st">    </span><span class="kw">left_join</span>(observed.tidy, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Sample"</span>, <span class="st">"Taxon"</span>))</span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(joint)</span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co">#&gt; # A tibble: 6 x 4</span></span>
<span id="cb11-9"><a href="#cb11-9"></a><span class="co">#&gt;   Sample Taxon Actual Observed</span></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="co">#&gt; 1 s1-23  Ava    0.143      351</span></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co">#&gt; 2 s1-35  Ava    0            1</span></span>
<span id="cb11-13"><a href="#cb11-13"></a><span class="co">#&gt; 3 s2-14  Ava    0.143      511</span></span>
<span id="cb11-14"><a href="#cb11-14"></a><span class="co">#&gt; 4 s2-4   Ava    0.5       1449</span></span>
<span id="cb11-15"><a href="#cb11-15"></a><span class="co">#&gt; 5 s2-5   Ava    0.5        883</span></span>
<span id="cb11-16"><a href="#cb11-16"></a><span class="co">#&gt; 6 s2-8   Ava    0            0</span></span></code></pre></div>
<p>The <code>left_join</code> operation only keeps taxa that are in <code>actual</code>, which in this case is just the 7 mock species.</p>
<p>Before proceeding, we will add a pseudocount to the observed read counts and filter reads from taxa that are not actually in the control samples (to avoid <code>Inf</code>s and <code>0</code>s in our error matrix, as discussed above),</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>joint &lt;-<span class="st"> </span>joint <span class="op">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Observed0 =</span> (Observed <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>) <span class="op">*</span><span class="st"> </span>(Actual <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>))</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(joint)</span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#&gt; # A tibble: 6 x 5</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="co">#&gt;   Sample Taxon Actual Observed Observed0</span></span>
<span id="cb12-6"><a href="#cb12-6"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">#&gt; 1 s1-23  Ava    0.143      351      352.</span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">#&gt; 2 s1-35  Ava    0            1        0 </span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="co">#&gt; 3 s2-14  Ava    0.143      511      512.</span></span>
<span id="cb12-10"><a href="#cb12-10"></a><span class="co">#&gt; 4 s2-4   Ava    0.5       1449     1450.</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">#&gt; 5 s2-5   Ava    0.5        883      884.</span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="co">#&gt; 6 s2-8   Ava    0            0        0</span></span></code></pre></div>
<p>Next, we compute the compositional error by dividing the observed abundance by the actual abundance,</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>joint &lt;-<span class="st"> </span>joint <span class="op">%&gt;%</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Error =</span> Observed0 <span class="op">/</span><span class="st"> </span>Actual)</span></code></pre></div>
<p>Because we treat the abundances and the error as compositional vectors, there is no need to normalize the abundances to proportions before or after computing the error. Finally, we get a matrix of the errors using the <code><a href="../reference/build_matrix.html">build_matrix()</a></code> helper function,</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>error_mat0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/build_matrix.html">build_matrix</a></span>(joint, Sample, Taxon, Error)</span></code></pre></div>
<p>This error matrix is the same we obtained by Approach 1, except that the row and column orders may differ.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all.equal">all.equal</a></span>(error_mat, error_mat0[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(error_mat), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(error_mat)])</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
</div>
<div id="estimate-the-bias-with-standard-errors" class="section level2">
<h2 class="hasAnchor">
<a href="#estimate-the-bias-with-standard-errors" class="anchor"></a>Estimate the bias with standard errors</h2>
<p>Next, we estimate the bias as the compositional mean, or <em>center</em>, of the errors in the control samples. We do this with the <code><a href="../reference/center.html">center()</a></code> function, which computes the center of a set of compositional vectors that are stored as a matrix. By default, <code><a href="../reference/center.html">center()</a></code> returns a named vector,</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="kw"><a href="../reference/center.html">center</a></span>(error_mat)</span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="co">#&gt;       Ava       Gva       Lcr       Lin       Pbi       Sam       Sag </span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="co">#&gt; 0.3423578 0.1789953 2.2954367 4.5106106 1.6271050 4.0254117 0.2406311</span></span></code></pre></div>
<p>For many purposes it is more useful to obtain a data frame,</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>bias &lt;-<span class="st"> </span><span class="kw"><a href="../reference/center.html">center</a></span>(error_mat, <span class="dt">enframe =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="st">    </span><span class="kw">rename</span>(<span class="dt">Bhat =</span> Center)</span>
<span id="cb17-3"><a href="#cb17-3"></a>bias</span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#&gt; # A tibble: 7 x 2</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#&gt;   Taxon  Bhat</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt;</span></span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="co">#&gt; 1 Ava   0.342</span></span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="co">#&gt; 2 Gva   0.179</span></span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="co">#&gt; 3 Lcr   2.30 </span></span>
<span id="cb17-10"><a href="#cb17-10"></a><span class="co">#&gt; 4 Lin   4.51 </span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="co">#&gt; 5 Pbi   1.63 </span></span>
<span id="cb17-12"><a href="#cb17-12"></a><span class="co">#&gt; 6 Sam   4.03 </span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="co">#&gt; 7 Sag   0.241</span></span></code></pre></div>
<p>We can get the pairwise bias with <code><a href="../reference/compute_ratios.html">compute_ratios()</a></code>,</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>bias.pw &lt;-<span class="st"> </span>bias <span class="op">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">    </span><span class="kw"><a href="../reference/compute_ratios.html">compute_ratios</a></span>(<span class="dt">group_vars =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>()) <span class="op">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Pair =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(Taxon.x, Taxon.y, <span class="dt">sep =</span> <span class="st">":"</span>))</span>
<span id="cb18-4"><a href="#cb18-4"></a>bias.pw</span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co">#&gt; # A tibble: 49 x 4</span></span>
<span id="cb18-6"><a href="#cb18-6"></a><span class="co">#&gt;    Taxon.x Taxon.y   Bhat Pair   </span></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;  </span></span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="co">#&gt;  1 Ava     Ava     1      Ava:Ava</span></span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="co">#&gt;  2 Ava     Gva     1.91   Ava:Gva</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="co">#&gt;  3 Ava     Lcr     0.149  Ava:Lcr</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="co">#&gt;  4 Ava     Lin     0.0759 Ava:Lin</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="co">#&gt;  5 Ava     Pbi     0.210  Ava:Pbi</span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co">#&gt;  6 Ava     Sag     1.42   Ava:Sag</span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="co">#&gt;  7 Ava     Sam     0.0850 Ava:Sam</span></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co">#&gt;  8 Gva     Ava     0.523  Gva:Ava</span></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="co">#&gt;  9 Gva     Gva     1      Gva:Gva</span></span>
<span id="cb18-17"><a href="#cb18-17"></a><span class="co">#&gt; 10 Gva     Lcr     0.0780 Gva:Lcr</span></span>
<span id="cb18-18"><a href="#cb18-18"></a><span class="co">#&gt; # … with 39 more rows</span></span></code></pre></div>
<p>The <code>bootrep_center</code> function can be used to estimate uncertainty in the estimated bias,</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>bootreps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bootrep_center.html">bootrep_center</a></span>(error_mat) <span class="op">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2"></a><span class="st">    </span><span class="kw">rename</span>(<span class="dt">Bhat =</span> Center)</span>
<span id="cb19-3"><a href="#cb19-3"></a>bootreps.summary &lt;-<span class="st"> </span>bootreps <span class="op">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="st">    </span><span class="kw">group_by</span>(Taxon) <span class="op">%&gt;%</span></span>
<span id="cb19-5"><a href="#cb19-5"></a><span class="st">    </span><span class="kw">summarize</span>(<span class="dt">Gm_mean =</span> <span class="kw"><a href="../reference/gm_mean.html">gm_mean</a></span>(Bhat), <span class="dt">Gm_se =</span> <span class="kw"><a href="../reference/gm_sd.html">gm_sd</a></span>(Bhat))</span>
<span id="cb19-6"><a href="#cb19-6"></a>bias0 &lt;-<span class="st"> </span><span class="kw">left_join</span>(bias, bootreps.summary, <span class="dt">by =</span> <span class="st">"Taxon"</span>)</span>
<span id="cb19-7"><a href="#cb19-7"></a>bias0</span>
<span id="cb19-8"><a href="#cb19-8"></a><span class="co">#&gt; # A tibble: 7 x 4</span></span>
<span id="cb19-9"><a href="#cb19-9"></a><span class="co">#&gt;   Taxon  Bhat Gm_mean Gm_se</span></span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="co">#&gt; 1 Ava   0.342   0.340  1.08</span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="co">#&gt; 2 Gva   0.179   0.179  1.09</span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="co">#&gt; 3 Lcr   2.30    2.31   1.08</span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="co">#&gt; 4 Lin   4.51    4.50   1.05</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="co">#&gt; 5 Pbi   1.63    1.63   1.07</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="co">#&gt; 6 Sam   4.03    4.02   1.06</span></span>
<span id="cb19-17"><a href="#cb19-17"></a><span class="co">#&gt; 7 Sag   0.241   0.241  1.05</span></span></code></pre></div>
<p>By converting the bootstrap estimates to pairwise estimates, we can also get standard errors for the pairwise bias,</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>bootreps.pw &lt;-<span class="st"> </span>bootreps <span class="op">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="st">    </span><span class="kw"><a href="../reference/compute_ratios.html">compute_ratios</a></span>(<span class="dt">group_vars =</span> <span class="st">".id"</span>)</span>
<span id="cb20-3"><a href="#cb20-3"></a>summary.pw &lt;-<span class="st"> </span>bootreps.pw <span class="op">%&gt;%</span></span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="st">    </span><span class="kw">group_by</span>(Taxon.x, Taxon.y) <span class="op">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5"></a><span class="st">    </span><span class="kw">summarize</span>(<span class="dt">Gm_mean =</span> <span class="kw"><a href="../reference/gm_mean.html">gm_mean</a></span>(Bhat), <span class="dt">Gm_se =</span> <span class="kw"><a href="../reference/gm_sd.html">gm_sd</a></span>(Bhat))</span>
<span id="cb20-6"><a href="#cb20-6"></a>bias.pw0 &lt;-<span class="st"> </span><span class="kw">left_join</span>(bias.pw, summary.pw, <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Taxon.x"</span>, <span class="st">"Taxon.y"</span>))</span>
<span id="cb20-7"><a href="#cb20-7"></a>bias.pw0</span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="co">#&gt; # A tibble: 49 x 6</span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="co">#&gt;    Taxon.x Taxon.y   Bhat Pair    Gm_mean Gm_se</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="co">#&gt;  1 Ava     Ava     1      Ava:Ava  1       1   </span></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="co">#&gt;  2 Ava     Gva     1.91   Ava:Gva  1.90    1.08</span></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="co">#&gt;  3 Ava     Lcr     0.149  Ava:Lcr  0.147   1.16</span></span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="co">#&gt;  4 Ava     Lin     0.0759 Ava:Lin  0.0756  1.13</span></span>
<span id="cb20-15"><a href="#cb20-15"></a><span class="co">#&gt;  5 Ava     Pbi     0.210  Ava:Pbi  0.208   1.13</span></span>
<span id="cb20-16"><a href="#cb20-16"></a><span class="co">#&gt;  6 Ava     Sag     1.42   Ava:Sag  1.41    1.12</span></span>
<span id="cb20-17"><a href="#cb20-17"></a><span class="co">#&gt;  7 Ava     Sam     0.0850 Ava:Sam  0.0847  1.10</span></span>
<span id="cb20-18"><a href="#cb20-18"></a><span class="co">#&gt;  8 Gva     Ava     0.523  Gva:Ava  0.526   1.08</span></span>
<span id="cb20-19"><a href="#cb20-19"></a><span class="co">#&gt;  9 Gva     Gva     1      Gva:Gva  1       1   </span></span>
<span id="cb20-20"><a href="#cb20-20"></a><span class="co">#&gt; 10 Gva     Lcr     0.0780 Gva:Lcr  0.0776  1.18</span></span>
<span id="cb20-21"><a href="#cb20-21"></a><span class="co">#&gt; # … with 39 more rows</span></span></code></pre></div>
</div>
<div id="evaluate-the-bias-estimate-visually" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluate-the-bias-estimate-visually" class="anchor"></a>Evaluate the bias estimate visually</h2>
<p>Plotting the bias estimate against the observed errors is useful to get a sense of the strength of bias versus noise in the control measurements were and how precisely the bias is estimated. When the control samples vary in their actual compositions, such plots can also allow us to confirm that the bias remains consistent across samples with different compositions. Since bias acts consistently on taxon ratios, and different control samples can have different taxa, we plot the error in the ratios for each pair of taxa over the samples where that pair was observed.</p>
<p>First, we get a data frame with the observed errors using the <code><a href="../reference/compute_ratios.html">compute_ratios()</a></code> function, and a data frame with the predicted ratios from the pairwise bias estimates (with standard errors) we found earlier,</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>ratios &lt;-<span class="st"> </span>joint <span class="op">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="st">    </span>compute_ratios <span class="op">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Pair =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(Taxon.x, Taxon.y, <span class="dt">sep =</span> <span class="st">":"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/is.finite">is.nan</a></span>(Error), Taxon.x <span class="op">&lt;</span><span class="st"> </span>Taxon.y)</span>
<span id="cb21-5"><a href="#cb21-5"></a>ratios.pred &lt;-<span class="st"> </span>bias.pw0 <span class="op">%&gt;%</span></span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Pair =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(Taxon.x, Taxon.y, <span class="dt">sep =</span> <span class="st">":"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Taxon.x <span class="op">&lt;</span><span class="st"> </span>Taxon.y)</span></code></pre></div>
<p>Next, we plot the observed error along with the estimated bias (multiplied and divided by two geometric standard errors),</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw">ggplot</span>(ratios, <span class="kw">aes</span>(Pair, Error, <span class="dt">color =</span> Sample)) <span class="op">+</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">1</span>, <span class="dt">color =</span> <span class="st">"grey"</span>) <span class="op">+</span></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="st">    </span><span class="kw">geom_pointrange</span>(<span class="dt">data =</span> ratios.pred, <span class="kw">aes</span>(<span class="dt">y =</span> Bhat, </span>
<span id="cb22-4"><a href="#cb22-4"></a>            <span class="dt">ymin =</span> Bhat <span class="op">/</span><span class="st"> </span>Gm_se<span class="op">^</span><span class="dv">2</span>, <span class="dt">ymax =</span> Bhat <span class="op">*</span><span class="st"> </span>Gm_se<span class="op">^</span><span class="dv">2</span>), </span>
<span id="cb22-5"><a href="#cb22-5"></a>        <span class="dt">color =</span> <span class="st">"black"</span>) <span class="op">+</span></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="st">    </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="fl">0.2</span>) <span class="op">+</span></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="st">    </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb22-8"><a href="#cb22-8"></a><span class="st">    </span><span class="kw">coord_flip</span>()</span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
</div>
<div id="calibration" class="section level1">
<h1 class="hasAnchor">
<a href="#calibration" class="anchor"></a>Calibration</h1>
<p>Next, we use the bias estimated from the control samples to calibrate the relative abundances in the non-control samples, according to Equation (9) of the manuscript. To illustrate how to apply calibration when we were only able to estimate bias for a subset of the observed taxa, we treat the <code>Other</code> taxonomic category as if it were an 8th species that we observed in the target samples but that was not in the controls.</p>
<div id="approach-1-matrices-1" class="section level3">
<h3 class="hasAnchor">
<a href="#approach-1-matrices-1" class="anchor"></a>Approach 1: Matrices</h3>
<p>First, get a matrix version of the observed compositions with all samples, not just the controls;</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>observed_mat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_matrix.html">as_matrix</a></span>(observed, <span class="dt">rownames =</span> Sample)</span></code></pre></div>
<p>Next, get a vector of the estimated bias, ensuring that taxa are in the same order as in the observed matrix,</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>bias_vec &lt;-<span class="st"> </span>bias <span class="op">%&gt;%</span><span class="st"> </span>deframe</span>
<span id="cb24-2"><a href="#cb24-2"></a>bias_vec &lt;-<span class="st"> </span>bias_vec[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(observed_mat)]</span>
<span id="cb24-3"><a href="#cb24-3"></a>bias_vec</span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="co">#&gt;       Ava       Gva       Lcr       Lin       Pbi       Sam       Sag </span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="co">#&gt; 0.3423578 0.1789953 2.2954367 4.5106106 1.6271050 4.0254117 0.2406311 </span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="co">#&gt;      &lt;NA&gt; </span></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="co">#&gt;        NA</span></span></code></pre></div>
<p>Note that any taxa in <code>observed</code> but not in <code>bias</code> have a value of <code>NA</code> in <code>bias_vec</code>. Finally, get a matrix with the calibrated compositions by dividing the observed compositions by the bias,</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>calibrated_mat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sweep">sweep</a></span>(observed_mat, <span class="dv">2</span>, bias_vec, <span class="st">"/"</span>)</span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(calibrated_mat) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="dv">2</span>)</span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="co">#&gt;           Ava     Gva     Lcr     Lin      Pbi     Sam     Sag Other</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">#&gt; s1-1     2.92    5.59 5955.29    0.00  4636.46    0.25 6258.54    NA</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="co">#&gt; s1-10 3002.71    0.00 4491.52    0.44     0.61 3713.16    8.31    NA</span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="co">#&gt; s1-11    0.00 1011.20 1723.42 1684.47     3.69    0.50    0.00    NA</span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="co">#&gt; s1-12 4159.39    0.00    0.00    0.22 13341.49    1.74    0.00    NA</span></span>
<span id="cb25-8"><a href="#cb25-8"></a><span class="co">#&gt; s1-13    0.00    0.00 1393.63    0.00  1139.45 1614.99    0.00    NA</span></span>
<span id="cb25-9"><a href="#cb25-9"></a><span class="co">#&gt; s1-14    0.00    0.00 4072.43    0.44     0.00    0.75 6129.72    NA</span></span></code></pre></div>
<p>Any measured taxa for which bias was not estimated cannot be calibrated, and so have calibrated values of <code>NA</code>. We can compute calibrated proportions for the subcompositions on just the control taxa with</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>calibrated_mat0 &lt;-<span class="st"> </span>calibrated_mat[, control_taxa]</span>
<span id="cb26-2"><a href="#cb26-2"></a>totals &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(calibrated_mat0, <span class="dv">1</span>, sum)</span>
<span id="cb26-3"><a href="#cb26-3"></a>calibrated_mat.prop &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sweep">sweep</a></span>(calibrated_mat0, <span class="dv">1</span>, totals, <span class="st">"/"</span>)</span>
<span id="cb26-4"><a href="#cb26-4"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(calibrated_mat.prop) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(<span class="dv">3</span>)</span>
<span id="cb26-5"><a href="#cb26-5"></a><span class="co">#&gt;         Ava   Gva   Lcr   Lin   Pbi   Sam   Sag</span></span>
<span id="cb26-6"><a href="#cb26-6"></a><span class="co">#&gt; s1-1  0.000 0.000 0.353 0.000 0.275 0.000 0.371</span></span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="co">#&gt; s1-10 0.268 0.000 0.400 0.000 0.000 0.331 0.001</span></span>
<span id="cb26-8"><a href="#cb26-8"></a><span class="co">#&gt; s1-11 0.000 0.229 0.390 0.381 0.001 0.000 0.000</span></span>
<span id="cb26-9"><a href="#cb26-9"></a><span class="co">#&gt; s1-12 0.238 0.000 0.000 0.000 0.762 0.000 0.000</span></span>
<span id="cb26-10"><a href="#cb26-10"></a><span class="co">#&gt; s1-13 0.000 0.000 0.336 0.000 0.275 0.389 0.000</span></span>
<span id="cb26-11"><a href="#cb26-11"></a><span class="co">#&gt; s1-14 0.000 0.000 0.399 0.000 0.000 0.000 0.601</span></span></code></pre></div>
</div>
<div id="approach-2-tidy-data-frames-1" class="section level3">
<h3 class="hasAnchor">
<a href="#approach-2-tidy-data-frames-1" class="anchor"></a>Approach 2: Tidy data frames</h3>
<p>Get a tidy data frame with the observed compositions and the bias</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>cal &lt;-<span class="st"> </span>observed.tidy <span class="op">%&gt;%</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="st">    </span><span class="kw">left_join</span>(bias, <span class="dt">by =</span> <span class="st">"Taxon"</span>)</span>
<span id="cb27-3"><a href="#cb27-3"></a>cal <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Sample <span class="op">==</span><span class="st"> "s1-1"</span>)</span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="co">#&gt; # A tibble: 8 x 4</span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="co">#&gt;   Sample Taxon Observed   Bhat</span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span id="cb27-7"><a href="#cb27-7"></a><span class="co">#&gt; 1 s1-1   Ava          1  0.342</span></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="co">#&gt; 2 s1-1   Gva          1  0.179</span></span>
<span id="cb27-9"><a href="#cb27-9"></a><span class="co">#&gt; 3 s1-1   Lcr      13670  2.30 </span></span>
<span id="cb27-10"><a href="#cb27-10"></a><span class="co">#&gt; 4 s1-1   Lin          0  4.51 </span></span>
<span id="cb27-11"><a href="#cb27-11"></a><span class="co">#&gt; 5 s1-1   Pbi       7544  1.63 </span></span>
<span id="cb27-12"><a href="#cb27-12"></a><span class="co">#&gt; 6 s1-1   Sam          1  4.03 </span></span>
<span id="cb27-13"><a href="#cb27-13"></a><span class="co">#&gt; 7 s1-1   Sag       1506  0.241</span></span>
<span id="cb27-14"><a href="#cb27-14"></a><span class="co">#&gt; 8 s1-1   Other        7 NA</span></span></code></pre></div>
<p>Calibrate each sample by dividing the observed abundance by the bias,</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>cal &lt;-<span class="st"> </span>cal <span class="op">%&gt;%</span></span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Calibrated =</span> Observed <span class="op">/</span><span class="st"> </span>Bhat)</span>
<span id="cb28-3"><a href="#cb28-3"></a>cal <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Sample <span class="op">==</span><span class="st"> "s1-1"</span>)</span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="co">#&gt; # A tibble: 8 x 5</span></span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="co">#&gt;   Sample Taxon Observed   Bhat Calibrated</span></span>
<span id="cb28-6"><a href="#cb28-6"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co">#&gt; 1 s1-1   Ava          1  0.342      2.92 </span></span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="co">#&gt; 2 s1-1   Gva          1  0.179      5.59 </span></span>
<span id="cb28-9"><a href="#cb28-9"></a><span class="co">#&gt; 3 s1-1   Lcr      13670  2.30    5955.   </span></span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="co">#&gt; 4 s1-1   Lin          0  4.51       0    </span></span>
<span id="cb28-11"><a href="#cb28-11"></a><span class="co">#&gt; 5 s1-1   Pbi       7544  1.63    4636.   </span></span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co">#&gt; 6 s1-1   Sam          1  4.03       0.248</span></span>
<span id="cb28-13"><a href="#cb28-13"></a><span class="co">#&gt; 7 s1-1   Sag       1506  0.241   6259.   </span></span>
<span id="cb28-14"><a href="#cb28-14"></a><span class="co">#&gt; 8 s1-1   Other        7 NA         NA</span></span></code></pre></div>
<p>Any measured taxa for which bias was not estimated cannot be calibrated, and so have calibrated values of <code>NA</code>. We can compute calibrated proportions for the subcompositions on just the control taxa with</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>cal.prop &lt;-<span class="st"> </span>cal <span class="op">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2"></a><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Taxon <span class="op">%in%</span><span class="st"> </span>control_taxa) <span class="op">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3"></a><span class="st">    </span><span class="kw">group_by</span>(Sample) <span class="op">%&gt;%</span></span>
<span id="cb29-4"><a href="#cb29-4"></a><span class="st">    </span><span class="kw">mutate_at</span>(<span class="kw">vars</span>(Observed, Calibrated), <span class="op">~</span><span class="st"> </span>. <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(.))</span>
<span id="cb29-5"><a href="#cb29-5"></a>cal.prop <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Sample <span class="op">==</span><span class="st"> "s1-1"</span>)</span>
<span id="cb29-6"><a href="#cb29-6"></a><span class="co">#&gt; # A tibble: 7 x 5</span></span>
<span id="cb29-7"><a href="#cb29-7"></a><span class="co">#&gt; # Groups:   Sample [1]</span></span>
<span id="cb29-8"><a href="#cb29-8"></a><span class="co">#&gt;   Sample Taxon  Observed  Bhat Calibrated</span></span>
<span id="cb29-9"><a href="#cb29-9"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span id="cb29-10"><a href="#cb29-10"></a><span class="co">#&gt; 1 s1-1   Ava   0.0000440 0.342  0.000173 </span></span>
<span id="cb29-11"><a href="#cb29-11"></a><span class="co">#&gt; 2 s1-1   Gva   0.0000440 0.179  0.000331 </span></span>
<span id="cb29-12"><a href="#cb29-12"></a><span class="co">#&gt; 3 s1-1   Lcr   0.602     2.30   0.353    </span></span>
<span id="cb29-13"><a href="#cb29-13"></a><span class="co">#&gt; 4 s1-1   Lin   0         4.51   0        </span></span>
<span id="cb29-14"><a href="#cb29-14"></a><span class="co">#&gt; 5 s1-1   Pbi   0.332     1.63   0.275    </span></span>
<span id="cb29-15"><a href="#cb29-15"></a><span class="co">#&gt; 6 s1-1   Sam   0.0000440 4.03   0.0000147</span></span>
<span id="cb29-16"><a href="#cb29-16"></a><span class="co">#&gt; 7 s1-1   Sag   0.0663    0.241  0.371</span></span></code></pre></div>
</div>
<div id="visualize-proportions-before-and-after" class="section level3">
<h3 class="hasAnchor">
<a href="#visualize-proportions-before-and-after" class="anchor"></a>Visualize proportions before and after</h3>
<p>In the Brooks dataset, each sample was constructed to be an even mixture of the chosen subset of species. Let’s check the effects of calibration on the estimated proportiosn for three of the 3-species samples.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>samples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"s1-6"</span>, <span class="st">"s1-2"</span>, <span class="st">"s1-3"</span>)</span>
<span id="cb30-2"><a href="#cb30-2"></a>plot_df &lt;-<span class="st"> </span>cal.prop <span class="op">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3"></a><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(Sample <span class="op">%in%</span><span class="st"> </span>samples) <span class="op">%&gt;%</span></span>
<span id="cb30-4"><a href="#cb30-4"></a><span class="st">    </span><span class="kw">gather</span>(<span class="st">"Type"</span>, <span class="st">"Proportion"</span>, <span class="st">"Observed"</span>, <span class="st">"Calibrated"</span>) <span class="op">%&gt;%</span></span>
<span id="cb30-5"><a href="#cb30-5"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Type =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(Type, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Observed"</span>, <span class="st">"Calibrated"</span>)))</span>
<span id="cb30-6"><a href="#cb30-6"></a><span class="kw">ggplot</span>(plot_df, <span class="kw">aes</span>(Type, Proportion, <span class="dt">fill =</span> Taxon)) <span class="op">+</span></span>
<span id="cb30-7"><a href="#cb30-7"></a><span class="st">    </span><span class="kw">geom_col</span>() <span class="op">+</span></span>
<span id="cb30-8"><a href="#cb30-8"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Sample) <span class="op">+</span></span>
<span id="cb30-9"><a href="#cb30-9"></a><span class="st">    </span><span class="kw">scale_fill_brewer</span>(<span class="dt">type =</span> <span class="st">"qual"</span>)</span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<p>The plot shows that the calibrated proportions are much closer to the nominally correct proportions of (0.33, 0.33, 0.33). Also, note that the proportion of <em>Prevotella bivia</em> (Pbi) is correctly calibrated downward in Sample “s1-3” but upward in proportion in Sample “s1-6”.</p>
</div>
</div>
<div id="appendix-taxonomic-overlap-needed-for-a-fully-determined-bias-estimate" class="section level1">
<h1 class="hasAnchor">
<a href="#appendix-taxonomic-overlap-needed-for-a-fully-determined-bias-estimate" class="anchor"></a>Appendix: Taxonomic overlap needed for a fully-determined bias estimate</h1>
<p>To fully estimate the bias of all taxa requires each taxon to appear in at least one control sample <em>and</em> that there is sufficient taxonomic overlap among the controls. A detailed explanation of this idea is given in the Appendix 2 of the manuscript; however, the basic principle can be easily understood with a few examples.</p>
<div id="scenario-1-two-control-samples-with-non-overlapping-sets-of-taxa" class="section level3">
<h3 class="hasAnchor">
<a href="#scenario-1-two-control-samples-with-non-overlapping-sets-of-taxa" class="anchor"></a>Scenario 1: Two control samples with non-overlapping sets of taxa</h3>
<p>Suppose that rather than the 7 control samples we used earlier, we had only measured samples “s2-9” and “s2-4”.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>error_mat1 &lt;-<span class="st"> </span>error_mat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"s2-9"</span>, <span class="st">"s2-4"</span>), ]</span>
<span id="cb31-2"><a href="#cb31-2"></a>error_mat1</span>
<span id="cb31-3"><a href="#cb31-3"></a><span class="co">#&gt;       Ava    Gva     Lcr Lin     Pbi   Sam Sag</span></span>
<span id="cb31-4"><a href="#cb31-4"></a><span class="co">#&gt; s2-9  NaN 1474.5 29617.5 NaN 19231.5   NaN NaN</span></span>
<span id="cb31-5"><a href="#cb31-5"></a><span class="co">#&gt; s2-4 2899    NaN     NaN NaN     NaN 41273 NaN</span></span></code></pre></div>
<p>Note that there are two taxa, “Lin” and “Sag”, that do not appear in either control sample, and so we won’t be able to estimate their efficiencies relative to the other taxa. However, we also have a problem if we wish to be able to estimate the bias of the 5 remaining taxa, due to the lack of taxonomic overlap between the two samples,</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="co"># List the taxa present in each sample</span></span>
<span id="cb32-2"><a href="#cb32-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(error_mat1, <span class="dv">1</span>, <span class="cf">function</span>(x) x[<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/is.finite">is.nan</a></span>(x)] <span class="op">%&gt;%</span><span class="st"> </span>names)</span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="co">#&gt; $`s2-9`</span></span>
<span id="cb32-4"><a href="#cb32-4"></a><span class="co">#&gt; [1] "Gva" "Lcr" "Pbi"</span></span>
<span id="cb32-5"><a href="#cb32-5"></a><span class="co">#&gt; </span></span>
<span id="cb32-6"><a href="#cb32-6"></a><span class="co">#&gt; $`s2-4`</span></span>
<span id="cb32-7"><a href="#cb32-7"></a><span class="co">#&gt; [1] "Ava" "Sam"</span></span></code></pre></div>
<p>For a given control sample, it is only possible to measure the relative efficiencies between the taxa in that sample, and we have no way to measure the relative efficiencies between the two groups of taxa in this case. There are infinitely many possible estimates of the bias that are equally consistent with the data, which maintain the ratios between taxa in the two groups but have different ratios between the groups. As a result, the <code><a href="../reference/center.html">center()</a></code> function will give an error when called with its default arguments in this case,</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="kw"><a href="../reference/center.html">center</a></span>(error_mat1)</span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="co">#&gt; Error in center(error_mat1): The center is not fully determined</span></span></code></pre></div>
<p>However, because it is still useful to be able to estimate the bias within each group (or component), the <code><a href="../reference/center.html">center()</a></code> function takes an argument, <code>components</code>, that lets us get a bias estimate anyways.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a><span class="kw"><a href="../reference/center.html">center</a></span>(error_mat1, <span class="dt">components =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb34-2"><a href="#cb34-2"></a><span class="st">    </span><span class="kw">arrange</span>(Component)</span>
<span id="cb34-3"><a href="#cb34-3"></a><span class="co">#&gt; # A tibble: 7 x 3</span></span>
<span id="cb34-4"><a href="#cb34-4"></a><span class="co">#&gt;   Taxon Center Component</span></span>
<span id="cb34-5"><a href="#cb34-5"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span id="cb34-6"><a href="#cb34-6"></a><span class="co">#&gt; 1 Ava    0.265         1</span></span>
<span id="cb34-7"><a href="#cb34-7"></a><span class="co">#&gt; 2 Sam    3.77          1</span></span>
<span id="cb34-8"><a href="#cb34-8"></a><span class="co">#&gt; 3 Gva    0.156         2</span></span>
<span id="cb34-9"><a href="#cb34-9"></a><span class="co">#&gt; 4 Lcr    3.14          2</span></span>
<span id="cb34-10"><a href="#cb34-10"></a><span class="co">#&gt; 5 Pbi    2.04          2</span></span>
<span id="cb34-11"><a href="#cb34-11"></a><span class="co">#&gt; 6 Lin    1             3</span></span>
<span id="cb34-12"><a href="#cb34-12"></a><span class="co">#&gt; 7 Sag    1             4</span></span></code></pre></div>
<p>The <code>Component</code> column gives the component of the taxon co-occurrence graph, which indicates the groups of taxa within which the bias was estimated.</p>
<p>To help examine the co-occurrence structure ahead of time, we offer several helper functions with the form <code>cooccurrence_*()</code>. In particular, you can view the co-occurrence network with</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cooccurrence.html">cooccurrence_network</a></span>(error_mat1)</span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(g, <span class="dt">edge.label=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(igraph<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/igraph/topics/E">E</a></span>(g)<span class="op">$</span>weight, <span class="dv">3</span>))</span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>The edge labels indicate the number of samples in which the pair of taxa co-occur.</p>
</div>
<div id="secario-2-a-third-control-sample-is-added-that-links-the-first-two" class="section level3">
<h3 class="hasAnchor">
<a href="#secario-2-a-third-control-sample-is-added-that-links-the-first-two" class="anchor"></a>Secario 2: A third control sample is added that links the first two</h3>
<p>Now suppose that we had also measured sample “s1-35” in addition to “s2-9” and “s2-4”.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>error_mat2 &lt;-<span class="st"> </span>error_mat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"s2-9"</span>, <span class="st">"s2-4"</span>, <span class="st">"s1-35"</span>), ]</span>
<span id="cb36-2"><a href="#cb36-2"></a>error_mat2</span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="co">#&gt;        Ava    Gva     Lcr   Lin     Pbi   Sam Sag</span></span>
<span id="cb36-4"><a href="#cb36-4"></a><span class="co">#&gt; s2-9   NaN 1474.5 29617.5   NaN 19231.5   NaN NaN</span></span>
<span id="cb36-5"><a href="#cb36-5"></a><span class="co">#&gt; s2-4  2899    NaN     NaN   NaN     NaN 41273 NaN</span></span>
<span id="cb36-6"><a href="#cb36-6"></a><span class="co">#&gt; s1-35  NaN 1070.0     NaN 33318 12762.0 21782 NaN</span></span>
<span id="cb36-7"><a href="#cb36-7"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(error_mat2, <span class="dv">1</span>, <span class="cf">function</span>(x) x[<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/is.finite">is.nan</a></span>(x)] <span class="op">%&gt;%</span><span class="st"> </span>names)</span>
<span id="cb36-8"><a href="#cb36-8"></a><span class="co">#&gt; $`s2-9`</span></span>
<span id="cb36-9"><a href="#cb36-9"></a><span class="co">#&gt; [1] "Gva" "Lcr" "Pbi"</span></span>
<span id="cb36-10"><a href="#cb36-10"></a><span class="co">#&gt; </span></span>
<span id="cb36-11"><a href="#cb36-11"></a><span class="co">#&gt; $`s2-4`</span></span>
<span id="cb36-12"><a href="#cb36-12"></a><span class="co">#&gt; [1] "Ava" "Sam"</span></span>
<span id="cb36-13"><a href="#cb36-13"></a><span class="co">#&gt; </span></span>
<span id="cb36-14"><a href="#cb36-14"></a><span class="co">#&gt; $`s1-35`</span></span>
<span id="cb36-15"><a href="#cb36-15"></a><span class="co">#&gt; [1] "Gva" "Lin" "Pbi" "Sam"</span></span></code></pre></div>
<p>This sample contains one of the previously missing taxa, “Lin”. More interestingly, it contains at least one taxon from each group of taxa contained in first two samples. This information linking the bias between the two groups allows us to obtain an estimate of the bias for all 6 taxa in the three controls, which we can confirm from the co-occurrence network,</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>g &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cooccurrence.html">cooccurrence_network</a></span>(error_mat2)</span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(g, <span class="dt">edge.label =</span> igraph<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/igraph/topics/E">E</a></span>(g)<span class="op">$</span>weight)</span></code></pre></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>Because the 7th taxon is still missing, we still need to use the <code>components = TRUE</code> option when estimating the bias,</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="kw"><a href="../reference/center.html">center</a></span>(error_mat2, <span class="dt">components =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-2"><a href="#cb38-2"></a><span class="st">    </span><span class="kw">arrange</span>(Component)</span>
<span id="cb38-3"><a href="#cb38-3"></a><span class="co">#&gt; # A tibble: 7 x 3</span></span>
<span id="cb38-4"><a href="#cb38-4"></a><span class="co">#&gt;   Taxon Center Component</span></span>
<span id="cb38-5"><a href="#cb38-5"></a><span class="co">#&gt;   &lt;chr&gt;  &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="co">#&gt; 1 Ava    0.186         1</span></span>
<span id="cb38-7"><a href="#cb38-7"></a><span class="co">#&gt; 2 Gva    0.127         1</span></span>
<span id="cb38-8"><a href="#cb38-8"></a><span class="co">#&gt; 3 Lcr    2.50          1</span></span>
<span id="cb38-9"><a href="#cb38-9"></a><span class="co">#&gt; 4 Lin    4.05          1</span></span>
<span id="cb38-10"><a href="#cb38-10"></a><span class="co">#&gt; 5 Pbi    1.58          1</span></span>
<span id="cb38-11"><a href="#cb38-11"></a><span class="co">#&gt; 6 Sam    2.65          1</span></span>
<span id="cb38-12"><a href="#cb38-12"></a><span class="co">#&gt; 7 Sag    1             2</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-the-demonstration-dataset">Load the demonstration dataset</a></li>
      <li>
<a href="#estimate-bias">Estimate bias</a><ul class="nav nav-pills nav-stacked">
<li><a href="#compute-the-error-matrix">Compute the error matrix</a></li>
      <li><a href="#estimate-the-bias-with-standard-errors">Estimate the bias with standard errors</a></li>
      <li><a href="#evaluate-the-bias-estimate-visually">Evaluate the bias estimate visually</a></li>
      </ul>
</li>
      <li><a href="#calibration">Calibration</a></li>
      <li><a href="#appendix-taxonomic-overlap-needed-for-a-fully-determined-bias-estimate">Appendix: Taxonomic overlap needed for a fully-determined bias estimate</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael McLaren.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
